versions/development/SPEC.md

```diff
@@ -2327,9 +2327,17 @@ Then the command would be instantiated as:
 /bin/mycmd /path/to/c.txt
 ```
 
-## Prepending a String to an Optional Parameter
+In general, it's invalid to supply an expression of optional type `T?` for a non-optional input or variable of type `T`, as the latter cannot accept `None`. (The idiom `select_first([expr, default])` coerces `expr : T?` to `T` by substituting `default : T` when `expr` is undefined; if `default` is omitted, then the coercion fails at runtime.) Similarly, an `Array[T]` expression does not satisfy an `Array[T]+` input.
 
-Sometimes, optional parameters need a string prefix.  Consider this task:
+These constraints also apply to function arguments, with exceptions for string concatenation, described below, and the equality/inequality operators `==` and `!=`, which can compare two values of types differing only in their quantifiers, considering `None` equal to itself but no other value.
+
+Quantifiers may apply within compound types; for example, if `a : Array[T?]+`, then `a[0] : T?`. An array type can be both non-empty and optional, `Array[T]+?`, such that the value can be `None` or a non-empty array, but not the empty array. (Reliance on this subtle distinction is discouraged, but it arises from other constructs such as a conditional call with an `Array[T]+` output.)
```

Comment:
Is the only valid form `Array[T]+?` or can it also be written `Array[T]?+`? The spec isn't explicit about this as far as I can see.

Reply:
I think that technically you can only apply `+` to arrays. Thus, since `Array[T]?` is not an array (it's an optional), the form `Array[T]?+` would be invalid